src/foam/nanos/auth/ruler/PreventDuplicateUsernameAction.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth.ruler',
+  name: 'PreventDuplicateUsernameAction',
+
+  documentation: `Prevents putting a user with an existing username.`,
+
+  implements: ['foam.nanos.ruler.RuleAction'],
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.X',
+    'foam.dao.DAO',
+    'foam.mlang.sink.Count',
+    'foam.nanos.auth.User',
+    'foam.util.SafetyUtil',
+    'net.nanopay.contacts.Contact',
+    'net.nanopay.model.Business',
+    'static foam.mlang.MLang.*'
+  ],
+
+  messages: [
+    { name: 'DUPLICATE_ERROR', message: 'User with same username already exists: ' }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        DAO userDAO = (DAO) x.get("localUserDAO");
+
+        if ( obj instanceof Business || obj instanceof Contact ) {
+          return;
+        }
+
+        User user = (User) obj;
+        if ( SafetyUtil.isEmpty(user.getUserName()) ) {
+          return;
```

Comment:
Is UserName required on create? If left blank we'll be returned a list of users with blank usernames when querying users using username. - Blank username will not be unique

Reply:
Will make username a requirement.